workflow/snakemake/Snakefile

```diff
@@ -0,0 +1,85 @@
+# Note that if you are working on the analysis development locally, i.e. outside
+# of the REANA platform, you can proceed as follows:
+#
+#   $ cd reana-demo-cms-h4l
+#   $ mkdir snakemake-local-run
+#   $ cd snakemake-local-run
+#   $ virtualenv ~/.virtualenvs/cms-h4l-snakemake
+#   $ source ~/.virtualenvs/cms-h4l-snakemake/bin/activate
+#   $ pip install snakemake
+#   $ cp -a ../code ../data .
+#   $ snakemake -s ../workflow/snakemake/Snakefile -p --cores 4 --use-singularity
+
+
+
+rule all:
+    input:
+        "results/mass4l_combine_userlvl3.pdf"
+
+
+rule analyze_data:
+    input:
+        "data",
+        "code"
+    output:
+        "results/DoubleMuParked2012C_10000_Higgs.root"
+    container:
+        "docker://cmsopendata/cmssw_5_3_32"
+    shell:
+        "source /opt/cms/cmsset_default.sh "
+        "&&  scramv1 project CMSSW CMSSW_5_3_32 "
+        "&& cd CMSSW_5_3_32/src "
+        "&& eval `scramv1 runtime -sh` "
+        "&& cp -r ../../code/HiggsExample20112012 . "
+        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
+        "&& scram b "
+        "&& cd ../Level3 "
+        "&& mkdir -p ../../../../results "
+        "&& cmsRun demoanalyzer_cfg_level3data.py"
+
+
+
+rule analyze_mc:
+    input:
+        "data",
+        "code"
+    output:
+        "results/Higgs4L1file.root"
+    container:
+        "docker://cmsopendata/cmssw_5_3_32"
+    shell:
+       "source /opt/cms/cmsset_default.sh "
+        "&&  scramv1 project CMSSW CMSSW_5_3_32 "
+        "&& cd CMSSW_5_3_32/src "
+        "&& eval `scramv1 runtime -sh` "
+        "&& cp -r ../../code/HiggsExample20112012 . "
+        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
+        "&& scram b "
+        "&& cd ../Level3 "
+        "&& mkdir -p ../../../../results "
+        "&& cmsRun demoanalyzer_cfg_level3MC.py"
+
+
+rule make_plot:
+    input:
+        "data",
+        "code",
+        mc_data="results/DoubleMuParked2012C_10000_Higgs.root",
+        a_data="results/Higgs4L1file.root"
+    output:
+        "results/mass4l_combine_userlvl3.pdf"
+    container:
+        "docker://cmsopendata/cmssw_5_3_32"
+    shell:
+        "source /opt/cms/cmsset_default.sh "
+        "&&  scramv1 project CMSSW CMSSW_5_3_32 "
+        "&& cd CMSSW_5_3_32/src "
+        "&& eval `scramv1 runtime -sh` "
+        "&& cp -r ../../code/HiggsExample20112012 . "
+        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
+        "&& scram b "
```

Comment:
FWIW I don't think it is necessary to repeat the scrum again here, since we are only doing the plots. We can test and simplify the workflow instructions it independently afterwards.

Reply:
Here is example of simplified Snakefile that runs scram as the first step. This also solves the concurrency problems, so one can use more cores; the example then runs in about 1min20sec for me.

```
rule all:
    input:
        "results/mass4l_combine_userlvl3.pdf"

rule scram:
    input:
        "data",
        "code"
    output:
        "results/scramdone.txt"
    container:
        "docker://cmsopendata/cmssw_5_3_32"
    shell:
        "source /opt/cms/cmsset_default.sh "
        "&&  scramv1 project CMSSW CMSSW_5_3_32 "
        "&& cd CMSSW_5_3_32/src "
        "&& eval `scramv1 runtime -sh` "
        "&& cp -r ../../code/HiggsExample20112012 . "
        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
        "&& scram b "
        "&& cd ../Level3 "
        "&& mkdir -p ../../../../results "
        "&& touch ../../../../results/scramdone.txt "

rule analyze_data:
    input:
        "code",
        "data",
        "results/scramdone.txt"
    output:
        "results/DoubleMuParked2012C_10000_Higgs.root"
    container:
        "docker://cmsopendata/cmssw_5_3_32"
    shell:
        "source /opt/cms/cmsset_default.sh "
        "&& cd CMSSW_5_3_32/src "
        "&& eval `scramv1 runtime -sh` "
        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
        "&& cd ../Level3 "
        "&& cmsRun demoanalyzer_cfg_level3data.py"

rule analyze_mc:
    input:
        "code",
        "data",
        "results/scramdone.txt"
    output:
        "results/Higgs4L1file.root"
    container:
        "docker://cmsopendata/cmssw_5_3_32"
    shell:
        "source /opt/cms/cmsset_default.sh "
        "&& cd CMSSW_5_3_32/src "
        "&& eval `scramv1 runtime -sh` "
        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
        "&& cd ../Level3 "
        "&& cmsRun demoanalyzer_cfg_level3MC.py"

rule make_plot:
    input:
        "code",
        "data",
        mc_data="results/DoubleMuParked2012C_10000_Higgs.root",
        a_data="results/Higgs4L1file.root"
    output:
        "results/mass4l_combine_userlvl3.pdf"
    container:
        "docker://cmsopendata/cmssw_5_3_32"
    shell:
        "source /opt/cms/cmsset_default.sh "
        "&& cd CMSSW_5_3_32/src "
        "&& eval `scramv1 runtime -sh` "
        "&& cd HiggsExample20112012/HiggsDemoAnalyzer "
        "&& cd ../Level3 "
        "&& root -b -l -q ./M4Lnormdatall_lvl3.cc"
```